File: apps/workspace-engine/pkg/reconcile/events/policyeval.go
package events

import (
	"context"

	"github.com/charmbracelet/log"
	"workspace-engine/pkg/reconcile"
)

const PolicyEvalKind = "policy-eval"

type PolicyEvalParams struct {
	WorkspaceID string
	VersionID   string
}

func EnqueuePolicyEval(
	queue reconcile.Queue,
	ctx context.Context,
	params PolicyEvalParams,
) error {
	return queue.Enqueue(ctx, reconcile.EnqueueParams{
		WorkspaceID: params.WorkspaceID,
		Kind:        PolicyEvalKind,
		ScopeType:   "deployment-version",
		ScopeID:     params.VersionID,
	})
}

func EnqueueManyPolicyEval(
	queue reconcile.Queue,
	ctx context.Context,
	params []PolicyEvalParams,
) error {
	if len(params) == 0 {
		return nil
	}
	log.Info("enqueueing policy eval", "count", len(params))
	items := make([]reconcile.EnqueueParams, len(params))
	for i, p := range params {
		items[i] = reconcile.EnqueueParams{
			WorkspaceID: p.WorkspaceID,
			Kind:        PolicyEvalKind,
			ScopeType:   "deployment-version",
			ScopeID:     p.VersionID,
		}
	}
	return queue.EnqueueMany(ctx, items)
}

File: packages/db/src/reconcilers/policy-eval.ts
import type { Tx } from "../common.js";
import type { ReconcileWorkScope } from "../schema/reconcile.js";
import { enqueue, enqueueMany } from "./enqueue.js";

const POLICY_EVAL_KIND = "policy-eval";

export async function enqueuePolicyEval(
  db: Tx,
  workspaceId: string,
  versionId: string,
): Promise<ReconcileWorkScope> {
  return enqueue(db, {
    workspaceId,
    kind: POLICY_EVAL_KIND,
    scopeType: "deployment-version",
    scopeId: versionId,
  });
}

export async function enqueueManyPolicyEval(
  db: Tx,
  items: Array<{ workspaceId: string; versionId: string }>,
): Promise<void> {
  return enqueueMany(
    db,
    items.map((item) => ({
      workspaceId: item.workspaceId,
      kind: POLICY_EVAL_KIND,
      scopeType: "deployment-version",
      scopeId: item.versionId,
    })),
  );
}
